File: problemList/slidingWindow/solution/LC2958.java
package problemList.slidingWindow.solution;

import java.util.HashMap;
import java.util.Map;

public class LC2958 {
    public int maxSubarrayLength(int[] nums, int k) {
        int n = nums.length, left = 0, ret = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for(int i = 0;i < n;i++){
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
            while(map.get(nums[i]) > k){
                map.put(nums[left], map.get(nums[left]) - 1);
                left++;
            }
            ret = Math.max(ret, i - left + 1);
        }
        return ret;
    }
}

File: SolutionsWithCorCpp/problem_list_0x3f/sliding_window/solution/LC2958.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

int maxSubarrayLength(vector<int> &nums, int k) {
    int n = nums.size(), ret = 0, r = 0;
    unordered_map<int, int> map;
    for(int i = 0;r < n;r++){
        map[nums[r]]++;
        while(map[nums[r]] > k && i < n){
            map[nums[i]]--;
            i++;
        }
        if(i <= r) ret = max(ret, r - i + 1);
    }
    return ret;
}